### PR TITLE
Adds first tests for client ser communication

### PR DESCRIPTION
**Pull request by priyapatil1**
@c-j-j
@dirv

### CODE REVIEW

File: src/main/java/ChatServer.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+        this.io = io;
+    }
+
+    public Socket makeConnection(ServerSocket serverSocket) {
+        try {
+            Socket socket = serverSocket.accept();
+            return socket;
+        } catch (IOException e) {
+            io.showOutput("Error in connecting socket");
+            e.printStackTrace();
```
Comment (c-j-j) [inline src/main/java/ChatServer.java]: You should never need to return null. What you can do here is `throw new UncheckedIOException(e)` on line 19, and just let it propagate up the stack. Logging an error and continuing is dangerous. Instead you should [fail fast](https://en.wikipedia.org/wiki/Fail-fast)

Comment (c-j-j) [inline src/main/java/ChatServer.java]: You should never need to return null. What you can do here is throw new UncheckedIOException(e) on line 19, and just let it propagate up the stack. Logging an error and continuing is dangerous. Instead you should fail fast

Comment (c-j-j) [inline src/main/java/ChatServer.java]: You should never need to return null. What you can do here is throw new UncheckedIOException(e) on line 19, and just let it propagate up the stack. Logging an error and continuing is dangerous. Instead you should fail fast

Comment (priyapatil1) [reply src/main/java/ChatServer.java]: 👍

**Comment by dirv** [comment]
You shouldn't check in the `target/` folder. Never check in anything which is build output. This goes for all programming languages.
Delete this folder using `git rm` and commit, and then add this folder to your .gitignore file and commit that. Use www.gitignore.io which does a good job of adding defaults.

File: src/main/java/App.java [hunk 2]
```diff
@@ -0,0 +1,19 @@
+import java.io.IOException;
+import java.net.InetAddress;
+import java.net.ServerSocket;
+import java.net.Socket;
+
+public class App {
+    public static void main(String[] args) throws IOException {
+        UserIO console = new UserIO(System.in, System.out);
```
Comment (dirv) [inline src/main/java/App.java:9]: This method is very busy. There are a lot of responsibilities hre. I like that you create a `UserIO` straight away. Why not do this for client and server apps?

File: src/main/java/ChatClient.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
+import java.io.*;
+import java.net.Socket;
+
+public class ChatClient {
+    private Socket socket;
```
Comment (dirv) [inline src/main/java/ChatClient.java]: I think you mean `private final`, surely?

File: src/main/java/ChatClient.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+        this.io = io;
+        this.socket = socket;
+    }
+
+    public void writeToServer() {
+        io.showInitialMessage(socket.getPort());
+        String message = io.getInput();
+        try {
+            PrintWriter printWriter = createPrintWriter();
+            writeMessageTillQuit(message, printWriter);
```
Comment (dirv) [inline src/main/java/ChatClient.java]: Til

File: src/main/java/ChatClient.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+        io.showExitMessage();
+        closeSocket();
+    }
+
+    private void writeMessageTillQuit(String name, PrintWriter printWriter) {
+        while (!name.contains("quit")) {
+            printWriter.println(name);
+            printWriter.flush();
+            try {
+                InputStream inputFromClient = socket.getInputStream();
```
Comment (dirv) [inline src/main/java/ChatClient.java]: This is quite a lot of boilerplate code - could you pull out the construction of the reader into a method?

File: src/main/java/ChatClient.java [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+                name = io.getInput();
+            } catch (IOException e) {
+                throw new UncheckedIOException(e);
+            }
+        }
+    }
+
+    private PrintWriter createPrintWriter() throws IOException {
+        OutputStream outToServer = socket.getOutputStream();
+        return new PrintWriter(outToServer, true);
```
Comment (dirv) [inline src/main/java/ChatClient.java]: What is this true? I would almost like to see a constant declared in here to describe what the parameter does.

File: src/main/java/ChatServer.java [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+
+    public ChatServer(ServerSocket serverSocket, UserIO io) {
+        this.serverSocket = serverSocket;
+        this.io = io;
+    }
+
+    public void readInFromClient() {
+        Socket server = makeConnection(serverSocket);
+        BufferedReader reader = createBufferedReader(server);
+        readInputTillOver(reader, server);
```
Comment (dirv) [inline src/main/java/ChatServer.java]: A till is a little drawer for money, also something that is done to land to prepare it for crops

File: src/main/java/ChatServer.java [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+        io.showExitMessage();
+        closeSocket();
+    }
+
+    private Socket makeConnection(ServerSocket serverSocket) {
+        try {
```
Comment (dirv) [inline src/main/java/ChatServer.java]: All these try-catch blocks make my eyes bleed.

Although you have the right idea, you can actually get away (I think) with having one try-catch on the `public` method. `private` methods are only ever called within the class so it's not as if they are leaking out to clients of this class. It's only the `public` methods that need a try-catch. Always try to put try-catch's at the highest level possible; when it comes to de-checking exceptions you can always put them at the public method.

File: src/main/java/UserIO.java [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+
+    public UserIO(InputStream input, PrintStream output) {
+        this.input = new BufferedReader(new InputStreamReader(input));
+        this.output = output;
+    }
+
+    public String getInput() {
+        String message = "";
+        try {
+            message = input.readLine();
```
Comment (dirv) [inline src/main/java/UserIO.java]: This is a super-long-winded way of saying `return input.readLine();`

File: src/test/java/ChatServerTest.java [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+    @Before public void setUp() throws IOException {
+        recordedOutput = new ByteArrayOutputStream();
+        output = new PrintStream(recordedOutput);
+        serverSocket = new ServerSocket(4444);
+        socket = new Socket("localhost", 4444);
+    }
+
+    @Test
+    public void makesAConnection() throws IOException {
+        UserIO console = createConsole("quit\n");
```
Comment (c-j-j) [inline src/test/java/ChatServerTest.java]: These tests feel like integration tests, where you are starting up the client and server, and testing the output of the server. There is value in having those tests, and it has resulted in you having a fairly simple design for your app as you're mainly using the Java IO libraries. However I would like to see more unit tests that don't require you to boot up the main parts of your app.

File: src/main/java/UserIO.java (hunk 9 above)
Comment (priyapatil1) [reply src/main/java/UserIO.java]: 👍

**Comment by c-j-j** [comment]
I noticed there's a WIP commit here. Do you know how you'd get rid of that/rename it to a proper commit message?

**Pull request closed without merging**